lib/Optimizer/Transforms/Mapping.cpp

```diff
@@ -671,11 +680,19 @@ struct Mapper : public cudaq::opt::impl::MappingPassBase<Mapper> {
     // and stop once you hit a used one. If you removed from the middle, you
     // would renumber the qubits, which would invalidate the mapping indices.
     for (unsigned i = sources.size() - 1; i >= numOrigQubits; i--) {
-      if (sources[i]->getUsers().empty())
+      if (sources[i]->use_empty())
         sources[i]->erase();
       else
         break;
     }
+    // Add sinks where needed
+    auto wireType = builder.getType<quake::WireType>();
+    builder.setInsertionPoint(block.getTerminator());
+    block.walk([&](Operation *op) {
+      for (auto r : op->getResults())
+        if (r.use_empty() && r.getType() == wireType)
+          builder.create<quake::SinkOp>(op->getLoc(), r);
+    });
```

Comment:
Another possibility would be to use the `phyToWire` vector in the router to place the sinks. This structure should hold the last wire corresponding to each qubit. This way, we wouldn't need to revisit the operation in the block. (I'm not sure if the performance gain would be substantial given the size of the circuits we currently deal with.)

Reply:
Good idea. Updated in c844c7fd.